File: examples/android/src/main/java/com/dropbox/core/examples/android/FilesAdapter.java
package com.dropbox.core.examples.android;

import android.content.Context;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.MimeTypeMap;
import android.widget.ImageView;
import android.widget.TextView;

import com.dropbox.core.v2.files.FileMetadata;
import com.dropbox.core.v2.files.FolderMetadata;
import com.dropbox.core.v2.files.Metadata;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Adapter for file list
 */
public class FilesAdapter extends RecyclerView.Adapter<FilesAdapter.MetadataViewHolder> {
    private List<Metadata> mFiles;
    private final Picasso mPicasso;
    private final Callback mCallback;

    public void setFiles(List<Metadata> files) {
        mFiles = Collections.unmodifiableList(new ArrayList<>(files));
        notifyDataSetChanged();
    }

    public interface Callback {
        void onFolderClicked(FolderMetadata folder);
        void onFileClicked(FileMetadata file);
    }

    public FilesAdapter(Picasso picasso, Callback callback) {
        mPicasso = picasso;
        mCallback = callback;
    }

    @Override
    public MetadataViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        Context context = viewGroup.getContext();
        View view = LayoutInflater.from(context)
                .inflate(R.layout.files_item, viewGroup, false);
        return new MetadataViewHolder(view);
    }

    @Override
    public void onBindViewHolder(MetadataViewHolder metadataViewHolder, int i) {
        metadataViewHolder.bind(mFiles.get(i));
    }

    @Override
    public long getItemId(int position) {
        return mFiles.get(position).getPathLower().hashCode();
    }

    @Override
    public int getItemCount() {
        return mFiles == null ? 0 : mFiles.size();
    }

    public class MetadataViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private final TextView mTextView;
        private final ImageView mImageView;
        private Metadata mItem;

        public MetadataViewHolder(View itemView) {
            super(itemView);
            mImageView = (ImageView)itemView.findViewById(R.id.image);
            mTextView = (TextView)itemView.findViewById(R.id.text);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {

            if (mItem instanceof FolderMetadata) {
                mCallback.onFolderClicked((FolderMetadata) mItem);
            }  else if (mItem instanceof FileMetadata) {
                mCallback.onFileClicked((FileMetadata)mItem);
            }
        }

        public void bind(Metadata item) {
            mItem = item;
            mTextView.setText(mItem.getName());

            // Load based on file path
            // Prepending a magic scheme to get it to
            // be picked up by DropboxPicassoRequestHandler

            if (item instanceof FileMetadata) {
                MimeTypeMap mime = MimeTypeMap.getSingleton();
                String ext = item.getName().substring(item.getName().indexOf(".") + 1);
                String type = mime.getMimeTypeFromExtension(ext);
                if (type != null && type.startsWith("image/")) {
                    mPicasso.load(FileThumbnailRequestHandler.buildPicassoUri((FileMetadata)item))
                            .placeholder(R.drawable.ic_photo_grey_600_36dp)
                            .error(R.drawable.ic_photo_grey_600_36dp)
                            .into(mImageView);
                } else {
                    mPicasso.load(R.drawable.ic_insert_drive_file_blue_36dp)
                            .noFade()
                            .into(mImageView);
                }
            } else if (item instanceof FolderMetadata) {
                mPicasso.load(R.drawable.ic_folder_blue_36dp)
                        .noFade()
                        .into(mImageView);
            }
        }
    }
}

File: examples/android/src/main/java/com/dropbox/core/examples/android/FilesAdapter.kt
package com.dropbox.core.examples.android

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.MimeTypeMap
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.dropbox.core.examples.android.FilesAdapter.MetadataViewHolder
import com.dropbox.core.v2.files.FileMetadata
import com.dropbox.core.v2.files.FolderMetadata
import com.dropbox.core.v2.files.Metadata
import com.squareup.picasso.Picasso
import java.util.Collections

/**
 * Adapter for file list
 */
class FilesAdapter(private val mPicasso: Picasso, private val mCallback: Callback) :
    RecyclerView.Adapter<MetadataViewHolder>() {
    private var mFiles: List<Metadata>? = null
    fun setFiles(files: List<Metadata>?) {
        mFiles = Collections.unmodifiableList(ArrayList(files))
        notifyDataSetChanged()
    }

    interface Callback {
        fun onFolderClicked(folder: FolderMetadata?)
        fun onFileClicked(file: FileMetadata?)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): MetadataViewHolder {
        val context = viewGroup.context
        val view = LayoutInflater.from(context)
            .inflate(R.layout.files_item, viewGroup, false)
        return MetadataViewHolder(view)
    }

    override fun onBindViewHolder(metadataViewHolder: MetadataViewHolder, i: Int) {
        metadataViewHolder.bind(mFiles!![i])
    }

    override fun getItemId(position: Int): Long {
        return mFiles!![position].pathLower.hashCode().toLong()
    }

    override fun getItemCount(): Int {
        return if (mFiles == null) 0 else mFiles!!.size
    }

    inner class MetadataViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        private val mTextView: TextView
        private val mImageView: ImageView
        private var mItem: Metadata? = null

        init {
            mImageView = itemView.findViewById<View>(R.id.image) as ImageView
            mTextView = itemView.findViewById<View>(R.id.text) as TextView
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            if (mItem is FolderMetadata) {
                mCallback.onFolderClicked(mItem as FolderMetadata?)
            } else if (mItem is FileMetadata) {
                mCallback.onFileClicked(mItem as FileMetadata?)
            }
        }

        fun bind(item: Metadata) {
            mItem = item
            mTextView.text = mItem!!.name

            // Load based on file path
            // Prepending a magic scheme to get it to
            // be picked up by DropboxPicassoRequestHandler
            if (item is FileMetadata) {
                val mime = MimeTypeMap.getSingleton()
                val ext = item.getName().substring(item.getName().indexOf(".") + 1)
                val type = mime.getMimeTypeFromExtension(ext)
                if (type != null && type.startsWith("image/")) {
                    mPicasso.load(FileThumbnailRequestHandler.Companion.buildPicassoUri(item))
                        .placeholder(R.drawable.ic_photo_grey_600_36dp)
                        .error(R.drawable.ic_photo_grey_600_36dp)
                        .into(mImageView)
                } else {
                    mPicasso.load(R.drawable.ic_insert_drive_file_blue_36dp)
                        .noFade()
                        .into(mImageView)
                }
            } else if (item is FolderMetadata) {
                mPicasso.load(R.drawable.ic_folder_blue_36dp)
                    .noFade()
                    .into(mImageView)
            }
        }
    }
}